File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/bulk/BulkRequest.java
package fr.myprysm.vertx.elasticsearch.action.bulk;

import fr.myprysm.vertx.elasticsearch.action.BaseRequest;
import fr.myprysm.vertx.elasticsearch.action.support.DocWriteRequest;
import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.experimental.Accessors;
import org.elasticsearch.action.bulk.BulkShardRequest;
import org.elasticsearch.action.support.WriteRequest;

import java.util.ArrayList;
import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = true)
@DataObject(generateConverter = true)
public class BulkRequest extends BaseRequest {

    private List<DocWriteRequest> requests;
    private WriteRequest.RefreshPolicy refreshPolicy;
    private long timeout = BulkShardRequest.DEFAULT_TIMEOUT.millis();
    private Integer waitForActiveShards;


    public BulkRequest add(DocWriteRequest request) {
        safeRequests().add(request);
        return this;
    }

    private List<DocWriteRequest> safeRequests() {
        if (requests == null) {
            requests = new ArrayList<>();
        }

        return requests;
    }

    public List<DocWriteRequest> getRequests() {
        return safeRequests();
    }

    /**
     * Build a new <code>BulkRequest</code> from a <code>JsonObject</code>,
     * calling parent constructor.
     *
     * @param json the <code>JsonObject</code>
     */
    public BulkRequest(JsonObject json) {
        super(json);
        BulkRequestConverter.fromJson(json, this);
    }

    /**
     * Transforms the <code>BulkRequest</code> into a <code>JsonObject</code>,
     * calling parent <code>toJson</code>.
     *
     * @return the <code>JsonObject</code>
     */
    public JsonObject toJson() {
        JsonObject json = super.toJson();
        BulkRequestConverter.toJson(this, json);
        return json;
    }

}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/bulk/BulkRequest.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.bulk

import fr.myprysm.vertx.elasticsearch.action.bulk.BulkRequest
import fr.myprysm.vertx.elasticsearch.action.support.DocWriteRequest
import org.elasticsearch.action.support.WriteRequest.RefreshPolicy

fun BulkRequest(
  headers: Map<String, String>? = null,
  refreshPolicy: RefreshPolicy? = null,
  requests: Iterable<fr.myprysm.vertx.elasticsearch.action.support.DocWriteRequest>? = null,
  timeout: Long? = null,
  waitForActiveShards: Int? = null): BulkRequest = fr.myprysm.vertx.elasticsearch.action.bulk.BulkRequest().apply {

  if (headers != null) {
    this.setHeaders(headers)
  }
  if (refreshPolicy != null) {
    this.setRefreshPolicy(refreshPolicy)
  }
  if (requests != null) {
    this.setRequests(requests.toList())
  }
  if (timeout != null) {
    this.setTimeout(timeout)
  }
  if (waitForActiveShards != null) {
    this.setWaitForActiveShards(waitForActiveShards)
  }
}

